simphony/io/serialisation.py

```diff
@@ -0,0 +1,290 @@
+                        value += '*' + str(refnum)
+
+                        # Check from _CUDSREFMAP dictionary if the object
+                        # referenced is not already saved
+                        if _CUDSREFMAP[item.uid] is not 'saved' or []:
+                            _CUDSREFMAP[item.uid] = 'saved'
+                            # Generate Yaml script for this component
+                            stream = _CUDSComponent_to_yaml(item, stream)
+                value += ']'
+
+        # Only one CUDSComponent?
+        elif isinstance(cdata[key], CUDSComponent):
+            refnum = _CUDSREFMAP.keys().index(cdata[key].uid) + 1
+            value = '*' + str(refnum)
+
+            # Use _CUDSREFMAP dictionary to mark already saved
+            # referenced objects
+            if _CUDSREFMAP[cdata[key].uid] is not 'saved' or []:
+                _CUDSREFMAP[cdata[key].uid] = 'saved'
+                # Generate Yaml script for this component
+                stream = _CUDSComponent_to_yaml(cdata[key], stream)
+
+        # Any of the simple data types?
+        elif type(cdata[key]) == str:
+            value = '"' + str(cdata[key]) + '"'
+        elif type(cdata[key]) == numpy.ndarray:
+            value = str(cdata[key].tolist())
+        elif cdata[key] is None:
+            value = None
+        else:
+            value = str(cdata[key])
+
+        # Write only key-value pairs where value is not None
+        if value:
+            yaml_comp += '    ' + str(key).replace('CUBA.', '') + \
+                ': ' + value + '\n'
+    stream = stream + yaml_comp
+    return stream
+
+
+def _dict_to_CUDSComponent(cubatype, comp, comp_dict={}):
+    """ Generate a CUDSComponent on the basis of a dictionary
+    provided by PyYaml library.
+
+    Parameters
+    ----------
+    cubatype: str
+        CUBA key (name) of the CUDSComponent in the format use in Yaml file
+    comp: dict
+        Dictionary object containing the parameters of the component.
+    comp_dict: dict
+        Dictionary where constructed CUDSComponents are added
+        with a Python id as reference: key-value pairs are of the format
+            id(CUDSComponent): CUDSComponent
+
+    """
+
+    # Check that comp does not refer to one of the components that
+    # have been already constructed
+    if id(comp) in comp_dict.keys():
+        return
+
+    if cubatype in KEYWORDS.keys():
+        # Find corresponding module and instantiate the class
+        mod = import_module('simphony.cuds.meta.%s' % cubatype.lower())
+        comp_class = getattr(mod, to_camel_case(cubatype))
+
+        # Must use (at least) two None parameters here, because currently
+        # there is no (easy) way to read required parameters for
+        # a class that is dynamically instantiated. A solution would be
+        # to define default values, e.g. 'None', for all parameters that
+        # do not have it defined yet.
+        comp_inst = comp_class(None, None)
```

Comment:
Since the meta data specifies certain fields to be "required", it makes sense to have mandatory arguments, otherwise the "required" attribute is only a description in the yaml file that never gets executed or presented in any way in the class.  Alternatively, it should be quite simple to add a property called "required_keys" to the generated class.
Related to how `comp_class` should be instantiated: The arguments in the generated `__init__` may not be the same each time the class is generated, as the yaml file is loaded into a dictionary whose order is lost.  Therefore I would suggest *always* instantiating a class with keyword arguments.  i.e. `Box(name='box1')` instead of `Box('box1')` because one day the first argument might be something else that is not `name`